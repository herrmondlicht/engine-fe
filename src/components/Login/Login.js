import React from "react";
import Button from "@material-ui/core/Button";
import { TextField, Grid, makeStyles, Paper } from "@material-ui/core";
import EngineImage from "./engine_original.PNG";

const useStyles = makeStyles(theme => ({
  root: {
    height: "100%",
    width: "100%",
    position: "absolute",
    background:
      "linear-gradient(22deg, rgba(131,58,180,1) 0%, rgba(252,176,69,1) 100%)"
    // background: "#E5E5E5"
  },
  loginContainer: {
    width: 350,
    padding: theme.spacing(3),
    boxSizing: "border-box"
  },
  field: {
    width: "100%"
  },
  engineImage: {
    width: "100%",
    padding: theme.spacing(2),
    boxSizing: "border-box"
  },
  loginButton: {
    marginTop: theme.spacing(2)
  }
}));

function Login() {
  const classes = useStyles();
  return (
    <Grid
      container
      justify="center"
      alignItems="center"
      className={classes.root}
    >
      <Paper className={classes.loginContainer}>
        <img src={EngineImage} className={classes.engineImage}></img>
        <LoginForm />
      </Paper>
    </Grid>
  );
}

function LoginForm() {
  const classes = useStyles();
  return (
    <Grid
      container
      item
      direction="column"
      spacing={2}
      alignItems="stretch"
      justify="center"
    >
      <Grid item>
        <TextField
          label="E-mail"
          className={classes.field}
          size="small"
          variant="outlined"
        ></TextField>
      </Grid>
      <Grid item>
        <TextField
          label="Senha"
          type="password"
          className={classes.field}
          size="small"
          variant="outlined"
        ></TextField>
      </Grid>
      <Grid item>
        <Button
          className={`${classes.loginButton} ${classes.field}`}
          variant="contained"
          color="primary"
        >
          Entrar
        </Button>
      </Grid>
    </Grid>
  );
}

export default Login;
