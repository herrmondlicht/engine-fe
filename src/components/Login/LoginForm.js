import React from "react";
import Button from "@material-ui/core/Button";
import { TextField, Grid, makeStyles } from "@material-ui/core";

const useStyles = makeStyles(theme => ({
  root: {
    height: "100%",
    width: "100%",
    position: "absolute",
    background: "rgba(0,0,0,0)"
  },
  loginButton: {
    marginTop: theme.spacing(2)
  },
  errorMessage: {
    width: "100%",
    textAlign: "center",
    color: theme.palette.error.main
  }
}));

function LoginForm({ userInput, changeInput, sendForm, errorMessage }) {
  const classes = useStyles();

  function onFormSubmit(e) {
    e.preventDefault();
    sendForm();
  }

  return (
    <form onSubmit={onFormSubmit} data-testid="LoginFormContainer">
      <Grid
        container
        item
        direction="column"
        spacing={2}
        alignItems="stretch"
        justify="center"
      >
        <Grid item>
          <TextField
            label="E-mail"
            data-testid="LoginFormContainer_Email"
            fullWidth
            size="small"
            variant="outlined"
            value={userInput.username}
            onChange={e =>
              changeInput({ ...userInput, username: e.target.value })
            }
          ></TextField>
        </Grid>
        <Grid item>
          <TextField
            label="Senha"
            data-testid="LoginFormContainer_Password"
            type="password"
            fullWidth
            size="small"
            value={userInput.password}
            variant="outlined"
            onChange={e =>
              changeInput({ ...userInput, password: e.target.value })
            }
          ></TextField>
        </Grid>
        <Grid item>
          <Button
            className={classes.loginButton}
            variant="contained"
            color="primary"
            type="submit"
            fullWidth
            data-testid="LoginForm_button"
          >
            Entrar
          </Button>
        </Grid>
        {errorMessage && (
          <Grid item container>
            <span className={`${classes.errorMessage}`}>{errorMessage}</span>
          </Grid>
        )}
      </Grid>
    </form>
  );
}

export default LoginForm;
