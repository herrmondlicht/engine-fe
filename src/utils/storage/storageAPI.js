export const storageAPI = (storage = window.localStorage) => ({
  setItem: (itemKey, value) => storage.setItem(itemKey, value),
  getItem: keyName => storage.getItem(keyName),
  removeItem: keyName => storage.removeItem(keyName)
});

export const STORAGE_KEYS = {
  TOKEN: "token"
};

export default storageAPI();
