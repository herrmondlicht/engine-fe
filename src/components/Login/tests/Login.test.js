import { render, fireEvent, wait } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createLogin } from "../Login";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { stub, assert, match } from "sinon";
import { STORAGE_KEYS } from "../../../utils/storage/storageAPI";

let apiRoutes = { login: { post: stub() } };
let storageAPI = { getItem: stub(), setItem: stub() };

const createTestComponent = props => {
  const Component = createLogin(apiRoutes, storageAPI);
  return render(<Component {...props} />, { wrapper: MemoryRouter });
};

describe("Login", () => {
  afterEach(() => {
    apiRoutes = { login: { post: stub() } };
    storageAPI = { getItem: stub(), setItem: stub() };
  });

  it("return of render", () => {
    const { asFragment } = createTestComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("return of render when token has already been provided", () => {
    storageAPI.getItem.withArgs(STORAGE_KEYS.TOKEN).returns("nonemptyValue");
    const { asFragment } = createTestComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("log the user in", async () => {
    const token = "tokenreturned";
    const username = "foolano";
    const password = "psw311";

    apiRoutes.login.post.resolves({ data: { token } });

    const { getByTestId, debug } = createTestComponent();

    userEvent.type(
      getByTestId("LoginFormContainer_Email").querySelector("input"),
      username
    );
    userEvent.type(
      getByTestId("LoginFormContainer_Password").querySelector("input"),
      password
    );
    fireEvent.click(getByTestId("LoginForm_button"));
    await wait(() => {
      assert.calledWith(
        apiRoutes.login.post,
        match({ data: { password, username } })
      );

      assert.calledWith(storageAPI.setItem, STORAGE_KEYS.TOKEN, token);
    });
  });
});
