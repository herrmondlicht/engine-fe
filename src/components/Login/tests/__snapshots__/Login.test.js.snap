// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login return of render 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      class="MuiPaper-root makeStyles-loginContainer-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <img
        alt="logo"
        class="makeStyles-engineImage-3"
        src="engine_logo.png"
      />
      <form
        data-testid="LoginFormContainer"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-direction-xs-column MuiGrid-justify-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              data-testid="LoginFormContainer_Email"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
                data-shrink="false"
              >
                E-mail
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-196 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-198"
                  >
                    <span>
                      E-mail
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              data-testid="LoginFormContainer_Password"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
                data-shrink="false"
              >
                Senha
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-196 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-198"
                  >
                    <span>
                      Senha
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-loginButton-136 MuiButton-containedPrimary MuiButton-fullWidth"
              data-testid="LoginForm_button"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Entrar
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Login return of render when token has already been provided 1`] = `<DocumentFragment />`;
